src/libawkward/Content.cpp

```diff
@@ -1189,81 +1189,48 @@ namespace awkward {
-    // getitem_next(missing.content()) in getitem_next(SliceMissing64) in a
-    // particular scenario; it can probably be generalized by handling more
-    // general scenarios.
-
-    if (that.get()->length() == 1  &&
-        dynamic_cast<SliceJagged64*>(missing.content().get())) {
-      ContentPtr tmp1 = that.get()->getitem_at_nowrap(0);
-      ContentPtr tmp2(nullptr);
-      if (IndexedOptionArray32* rawtmp1 =
-          dynamic_cast<IndexedOptionArray32*>(tmp1.get())) {
-        tmp2 = rawtmp1->project();
-        if (!check_missing_jagged_same(that, rawtmp1->bytemask(), missing)) {
-          return that;
-        }
-      }
-      else if (IndexedOptionArray64* rawtmp1 =
-               dynamic_cast<IndexedOptionArray64*>(tmp1.get())) {
-        tmp2 = rawtmp1->project();
-        if (!check_missing_jagged_same(that, rawtmp1->bytemask(), missing)) {
-          return that;
-        }
-      }
-      else if (ByteMaskedArray* rawtmp1 =
-               dynamic_cast<ByteMaskedArray*>(tmp1.get())) {
-        tmp2 = rawtmp1->project();
-        if (!check_missing_jagged_same(that, rawtmp1->bytemask(), missing)) {
-          return that;
-        }
-      }
-      else if (BitMaskedArray* rawtmp1 =
-               dynamic_cast<BitMaskedArray*>(tmp1.get())) {
-        tmp2 = rawtmp1->project();
-        if (!check_missing_jagged_same(that, rawtmp1->bytemask(), missing)) {
-          return that;
-        }
-      }
-
-      if (tmp2.get() != nullptr) {
-        return std::make_shared<RegularArray>(Identities::none(),
-                                              that.get()->parameters(),
-                                              tmp2,
-                                              tmp2.get()->length());
-      }
+  const ContentPtr getitem_next_missing_jagged(const SliceMissing64& missing,
+                                               const Slice& tail,
+                                               const Index64& advanced,
+                                               const ContentPtr& that) {
+    const SliceJagged64* jagged = dynamic_cast<SliceJagged64*>(missing.content().get());
+    if ( jagged == nullptr ) {
+      throw std::runtime_error("Logic error: calling getitem_next_missing_jagged with bad slice type");
+    }
+    const Index64 index = missing.index();
+    ContentPtr content = that.get()->getitem_at_nowrap(0);
+    if ( content.get()->length() != index.length() ) {
+      throw std::invalid_argument(
+        std::string("cannot fit masked jagged slice with length ")
+        + std::to_string(index.length()) + std::string(" into ")
+        + that.get()->classname() + std::string(" of size ") + std::to_string(content.get()->length()));
     }
-    return that;
+    Index64 outputmask(index.length());
+    Index64 starts(index.length());
+    Index64 stops(index.length());
+    Content_getitem_next_missing_jagged_getmaskstartstop(index.ptr().get() + index.offset(),
```

Comment:
Here's my example of incrementing the pointer before passing to C function. We could provide a convenience function in `Index` type to do that in one call as well. Any reason not to go this route and pass the offset into the kernel?

Reply:
Exactly: I'm catching up to you. That would be a good refactoring, but I think this should be done in a single sweep, one that I should probably do someday. However, that should be done all at once when no one else is working on the kernels (@reikdas and @trickarcher).

For right now, just make them fit the established pattern. The improvement should happen all at once.

Reply:
Are you saying I should put the offset into the kernel signature for this PR as well? Or can I leave it as-is?

Reply:
For the purposes of factorization, just add the offset as a separate function argument to fit the pattern. Then when you or I am converting them all in bulk (in a new PR), the deviation from the pattern won't be confusing.

I just noticed that you did that by simply adding to the pointer. I thought that would be a type error—I thought C++ strictly separated pointers and integers. I can see how it knows about the `sizeof` when adding—it's some overloaded operator. Does it only work in some version of the standard?

When you or I update these, I think it should be in a named method (`rawdata`?), so that it's easier to do future sweeps through the code. (Waking through by incremental text search is reliable if semantic constructs always have the same textual pattern. In the case of this sweep, the text to search is `pter().get()` because that occurs in all instances where we'll want to change and few others.)

Reply:
Ok. Regarding pointer + integer, it's C standard: http://c0x.coding-guidelines.com/6.5.6.html 1167
Although there is some nuance to whether the pointer actually points to an array.